File: DR_oldmatlabcode/Src/sj_Matlab/Matlab/loren_inpath/Adaptive/Suzuki/cardinal.m
% points = CARDINAL(cpx, cpy, npoints)
%          Retrns the x and y coordinates of npoints evenly spaced points on the spline
function [p] = cardinal(cpx, cpy, npoints)

t = 0;
s = (1 - t) / 2;
Mc = [(-s)  (2-s)    (s-2)   s;
     (2*s)  (s-3)    (3-2*s) (-s);
     (-s)     0        s      0;
     0        1        0       0];


p = zeros(npoints,2);
p(:,1) = (cpx(2):((cpx(end-1)-cpx(2))/(npoints-1)):cpx(end-1))';

% for each point, find the segment it is in
for i = 1:npoints
	% find the segment of the spline containing the current point %
	cseg = 2;
	while (cpx(cseg+1) < p(i,1))
		cseg = cseg + 1;
	end
	x1 = (p(i,1) - cpx(cseg)) / (cpx(cseg+1) - cpx(cseg));
	x2 = x1 * x1;
	x3 = x2 * x1;
	p(i,2) = (cpy(cseg-1)*(-.5*x3 + x2 -.5*x1) + ...
			   cpy(cseg)*(1.5*x3 - 2.5*x2 + 1) + ...
			   cpy(cseg+1)*(-1.5*x3 + 2.0*x2 + .5*x1) + ...
			   cpy(cseg+2)*(.5*x3 - .5*x2));  
end



%for i = 1:npoints
%	% find the segment of the spline containing the current point %
%	cseg = 2;
%	while (cpx(cseg+1) < p(i,1))
%		cseg = cseg + 1;
%	end
%	x1 = (p(i,1) - cpx(cseg)) / (cpx(cseg+1) - cpx(cseg));
%	x2 = x1 * x1;
%	x3 = x2 * x1;
%	p(i,2) = exp(cpy(cseg-1)*(-.5*x3 + x2 -.5*x1) + ...
%			   cpy(cseg)*(1.5*x3 - 2.5*x2 + 1) + ...
%			   cpy(cseg+1)*(-1.5*x3 + 2.0*x2 + .5*x1) + ...
%			   cpy(cseg+2)*(.5*x3 - .5*x2));  
%end

File: DR_oldmatlabcode/Src/sj_Matlab/Matlab/loren_inpath/Adaptive/Suzuki/GL/cardinal.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define TENSION0        /* for a cardinal spline tension of 0 */
/*#define TENSION1 */        /* for a cardinal spline tension of 1 */


void cardinal(float *cpx, float *cpy, int ncp, float *x, float *y, int npoints)
    /* for each of the npoints values in the array x, returns in y the value of the
     * cardinal spline at that point.  */
{
    int     i,j;
    float    x1, x2, x3;


    for (i = 0; i < npoints ; i++) {
        /* find the segment of the spline containing the current point */
        j = 1;
        while (((cpx[j] > x[i]) || (cpx[j+1] < x[i])) && (j <= ncp - 2)){
            j++;
        }
        if (j == ncp - 2) {
            fprintf(stderr, "Error in cardinal: point outside spline bounds\n");
            exit(1);
        }
        /* compute the value of the spline at this point */
        x1 = (float) (x[i] - cpx[j]) / (float) (cpx[j+1] - cpx[j]);
        x2 = x1 * x1;
        x3 = x2 * x1;
#ifdef TENSION0
        y[i] = (cpy[j-1]*(-.5*x3 + x2 -.5*x1) + cpy[j]*(1.5*x3 - 2.5*x2 + 1) + 
                   cpy[j+1]*(-1.5*x3 + 2.0*x2 + .5*x1) + cpy[j+2]*(.5*x3 - .5*x2));  
#endif
#ifdef TENSION1
        y[i] = (cpy[j]*(2*x3 - 3*x2 + 1) + cpy[j+1]*(-2*x3 + 3.0*x2));
#endif
        if (y[i] < 0) y[i] = 0;  
    }
}
